src/bp/core/modules/config-reader.ts

```diff
@@ -172,21 +182,56 @@ export class ConfigReader {
     return config
   }
 
-  @Memoize()
+  private async getCachedOrFresh(key, getFresh: { fn: (...args) => any; args: any[] }) {
+    let config
+    if (!this.moduleConfigCache.has(key)) {
+      config = await getFresh.fn.call(this, ...getFresh.args)
+      this.moduleConfigCache.set(key, config)
+    } else {
+      config = this.moduleConfigCache.get(key) || {}
+    }
+    return config
+  }
+
   public async getGlobal(moduleId: string): Promise<ModuleConfig> {
-    return this.getMerged(moduleId)
+    return this.getCachedOrFresh(moduleId, { fn: this.getMerged, args: [moduleId] })
   }
 
-  // Don't @Memoize() this fn. It only memoizes on the first argument
-  // https://github.com/steelsojka/lodash-decorators/blob/master/src/memoize.ts#L15
   public getForBot(moduleId: string, botId: string, ignoreGlobal?: boolean): Promise<ModuleConfig> {
     const cacheKey = `${moduleId}//${botId}//${!!ignoreGlobal}`
-    return this.getForBotMemoized(cacheKey)
+    return this.getCachedOrFresh(cacheKey, { fn: this.getMerged, args: [moduleId, botId, yn(!!ignoreGlobal)] })
   }
 
-  @Memoize()
-  public getForBotMemoized(cacheKey: string): Promise<ModuleConfig> {
-    const [moduleId, botId, ignoreGlobal] = cacheKey.split('//')
-    return this.getMerged(moduleId, botId, yn(ignoreGlobal))
+  private _listenForModuleConfigCacheInvalidation() {
+    async function deleteCacheStartingWith(prefix: string, cache: LRU<string, any>): Promise<void> {
+      const keys = cache.keys().filter(x => {
+        return x.startsWith(prefix)
+      })
+      keys.forEach(x => cache.del(x))
+    }
+
+    this.cache.events.on('invalidation', async key => {
+      try {
+        const isGlobalModuleConfig = key.indexOf('data/global/config') !== -1
+        const matches = !isGlobalModuleConfig
+          ? key.match(/^([A-Z0-9-_]+)::data\/bots\/([A-Z0-9-_]+)\/config\/([\s\S]+([A-Z0-9-_])+\.json)/i)
+          : key.match(/^([A-Z0-9-_]+)::data\/global\/config\/([\s\S]+([A-Z0-9-_])+\.json)/i)
+        const type = matches && matches.length >= 2 && matches[1]
+
+        if (type && (type === 'file' || type === 'object')) {
+          const moduleId = matches[isGlobalModuleConfig ? 2 : 3].replace('.json', '')
+
+          if (isGlobalModuleConfig) {
+            await deleteCacheStartingWith(moduleId, this.moduleConfigCache)
+          } else {
+            const botId = matches[2]
+            await deleteCacheStartingWith(`${moduleId}//${botId}//true`, this.moduleConfigCache)
+            await deleteCacheStartingWith(`${moduleId}//${botId}//false`, this.moduleConfigCache)
+          }
+        }
```

Comment:
This is a bit over-engineered. 
Module config isn't edited often, no need to be that specific when deleting the cache. Just delete every cache for that module. You can also use optional chaining to simplify it: 
```js
const moduleId = key.match(^.*::data\/.*\/config\/([\s\S]+([a-zA-Z0-9-_])+\.json)/i)?.[1]?.replace('.json', '')

if(moduleId){
   deleteCacheStartingWith(moduleId)
}

Reply:
Yes, my first thought was also to just delete everything for that module, but since changing config for one bot would delete the cache for another one, is was reluctant, I changed to be a single handler as requested

Reply:
Yes, my hesitation to do this came from the fact that changing local config in one bot will invalidate the cache in the other ones, changed to a single handler as requested.

Reply:
I agree, but the benefits are pretty small considering how often module config is changed and how expensive the operation to process it is, sometime simple is better, less edge cases to handle